RangeControl: Convert component to TypeScript

aaronrobertshaw opened the pull request:
Depends On:
- https://github.com/WordPress/gutenberg/pull/40518

Related:
- https://github.com/WordPress/gutenberg/issues/40507
- https://github.com/WordPress/gutenberg/pull/41444
- https://github.com/WordPress/gutenberg/pull/41445

## What?
Converts the `RangeControl` component to TypeScript. 

_(Unit tests and story TypeScript conversions are in separate PRs, #41444 & #41445)_

## Why?
While looking to create a new component combining both a `UnitControl` and `RangeControl` (https://github.com/WordPress/gutenberg/pull/40462) it presented an opportunity to improve the slider components we offered in a more composable way. 

The first step in this process was to type the `RangeControl` component.

See https://github.com/WordPress/gutenberg/issues/40507 for more details.

## How?
Refactors the current `RangeControl` to TypeScript.

## Testing Instructions
1. Run `npm run dev:package-types` and ensure no typing errors
2. Run `npm run test-unit packages/components/src/range-control/test/` and check they all pass
3. Fire up the Storybook and confirm the RangeControl continues to function as expected
4. Test the various range controls within the block/site editors in Gutenberg still work

ciampo commented on packages/components/src/range-control/index.tsx:
```diff
@@ -121,16 +129,16 @@ function RangeControl(
-	const handleOnChange = ( nextValue ) => {
-		nextValue = parseFloat( nextValue );
+	const handleOnChange = ( nextValue: number | string | null ) => {
+		nextValue = parseFloat( nextValue as string );
```
I wonder if this solution would be better

```suggestion
		nextValue =
			nextValue !==  null ? parseFloat( `${nextValue}` ) : nextValue;
```

Whatever approach we decide to pick here, the same should apply in other parts of the code (like in the `handleOnReset` callback)

aaronrobertshaw replied:
I've tried to tidy this up and make the `handleOnChange` and `handOnReset` more consistent. Let me know if this is sufficient or what else you had in mind. (https://github.com/WordPress/gutenberg/pull/40535/commits/b17df9a2bacc8075b6df7097678edee91514b371)

ciampo commented on packages/components/src/range-control/index.tsx:
```diff
@@ -220,11 +230,11 @@ function RangeControl(
 						describedBy={ describedBy }
 						disabled={ disabled }
-						id={ id }
+						id={ id as string }
```
Probably safer to make sure that `id` is an actual string?

```suggestion
						id={ `${id}` }
```

ciampo reviewed (commented):
Great job so far, @aaronrobertshaw !

I have a couple of observations / questions:

- Is there particular reason why  only some components are typed through `WordPressComponentProps` ? (`RangeMark`, `RangeRail`, `Marks` are not)

- Regarding the `value` prop:

  - we should consider if we should type it as `number | string`, and handle any parsing internally.
  - we should also try to align the `null` and `undefined` instances, and try to pick one out of the two

ciampo commented:
Also, flagging #40568 (and #40518) which contain some changes to `RangeControl` (bug fixes) and may be merged before this PR (and therefore require rebasing).

aaronrobertshaw commented:
I appreciate the early feedback @ciampo, thank you 👍 

> Is there particular reason why only some components are typed through WordPressComponentProps ? (RangeMark, RangeRail, Marks are not)

I don't believe there was any reason except that I missed `RangedRail` and `Marks`. `RangedMark` was already typed through `WordPressComponentProps` though.

> Regarding the value prop:
> 
> - we should consider if we should type it as number | string, and handle any parsing internally.
> - we should also try to align the null and undefined instances, and try to pick one out of the two

I think from the `RangeControl`'s perspective typing `value` as an optional `number` should probably be maintained. My understanding of the original code's intent was that it is responsible for handling parsing string, undefined or null values prior to calling the `onChange` callback. 

I'm probably missing something here so happy to take on board any further suggestions.

> Also, flagging https://github.com/WordPress/gutenberg/pull/40568 (and https://github.com/WordPress/gutenberg/pull/40518) which contain some changes to RangeControl (bug fixes) and may be merged before this PR (and therefore require rebasing).

Thanks for the heads up.

I've had a quick look at those PRs and it's probably best if they do land first given their importance and functional changes. I'll try and help out with reviews or testing on those, then resolve the ensuing conflicts on this PR.

ciampo commented:
> I don't believe there was any reason except that I missed `RangedRail` and `Marks`. `RangedMark` was already typed through `WordPressComponentProps` though.

Thank you! And you're right re. `RangedMark`, I must have gotten confused while writing my previous message.

> I think from the `RangeControl`'s perspective typing `value` as an optional `number` should probably be maintained. My understanding of the original code's intent was that it is responsible for handling parsing string, undefined or null values prior to calling the `onChange` callback. 

The point that you make is totally valid, and is actually the main point of debate around converting to TypeScript the `*Control` components dealing with numbers:

- `InputControl` accepts `value` as a `string` (I guess no objections here)
- `NumberControl` is not typed yet. I had a first attempt in #38753, but I realized that we'd probably need to untangle its "dependant" components first (e.g `UnitControl`, `RangeControl`). Here we could debate if `value` should be only a `number`, only a `string`, or a `number|string`.
- `UnitControl`'s value can currently be a `number|string` — which is understandable, since `UnitControl` accepts quantity and units (e.g. `14px`)
- `RangeControl`. I would argue that, as of today, `RangeControl` should behave the same way we'd like `NumberControl` to behave (see point above).

And so, by picking the `value` type for `RangeControl`, we are inherently picking it for `NumberControl`.

Here's a few considerations off the top of my head:

- I agree that the purest, cleanest way to deal with this is to expect `value` to be a `number`
- On the other hand, technically an `input` element expects `value` to be a `string`
- We also need to make sure that whatever we choose also works with the way the components have been used (i.e. we still need to make sure we fallback graciously in case, for example, we expect only a `number` but a `string` is passed)
- Same for `null` and `undefined` values.

Probably the best way to move forward, considering that `RangeControl` is _not_ marked as experimental, would be to keep the type of `value` as `number` for now, and potentially make any adjustments later if necessary.

We should also assess how `RangeControl` is currently used in the repo, to inform our future decisions better.

(curious to know that @mirka thinks about this)

(Coincidentally, we have [a related conversation](https://github.com/WordPress/gutenberg/pull/40633#discussion_r860883727) on the PR migrating `TextControl` to TypeScript)

> I've had a quick look at those PRs and it's probably best if they do land first given their importance and functional changes. I'll try and help out with reviews or testing on those, then resolve the ensuing conflicts on this PR.

Agreed, thank you for understanding and prioritizing accordingly!

mirka commented:
> Probably the best way to move forward, considering that `RangeControl` is _not_ marked as experimental, would be to keep the type of `value` as `number` for now, and potentially make any adjustments later if necessary.
>
> (curious to know that @mirka thinks about this)

This sounds reasonable to me. To be honest I don't have much useful thoughts to add here since I haven't actually fought this particular chain of dragons yet 😇

aaronrobertshaw commented:
I've updated the base branch for this PR to be https://github.com/WordPress/gutenberg/pull/40518 and resolved conflicts. I'll revisit it further once https://github.com/WordPress/gutenberg/pull/40518 lands.

aaronrobertshaw committed:
Convert RangeControl to TypeScript

aaronrobertshaw commented:
After a change in approach in #40518 I've rebased this and fixed up the typing again. 

I'm out of time today to confirm and test but this PR shouldn't need to depend on #40518 now given it no longer updates the `RangeControl` component. I'll clean this up tomorrow.

ciampo commented:
> After a change in approach in #40518 I've rebased this and fixed up the typing again.
> 
> I'm out of time today to confirm and test but this PR shouldn't need to depend on #40518 now given it no longer updates the `RangeControl` component. I'll clean this up tomorrow.

Thank you for the update! Let us know once this PR is ready for another round of review :)

aaronrobertshaw committed:
Add named export for docgen

aaronrobertshaw committed:
Add changelog

aaronrobertshaw committed:
Fix naming of component

stokesman commented on packages/components/src/range-control/types.ts:
```diff
@@ -0,0 +1,282 @@
+	value?: ControlledRangeValue
+) => string | number | null | undefined;
+
+export type RangeControlProps< P > = Pick<
```
Since `P` is only used by optional props should it also be optional?
```suggestion
export type RangeControlProps< P = unknown > = Pick<
```

ciampo replied:
Good point. I also think we could give it a more explicit name, maybe? Something like `IconProps` maybe ?

aaronrobertshaw replied:
Thanks for the suggestion 👍 

I've updated this in https://github.com/WordPress/gutenberg/pull/40535/commits/93c573e5becb4b0ee669547688b17afaf2361957.

aaronrobertshaw committed:
Rename and make icon props optional

aaronrobertshaw committed:
Add additional descriptions for icon types

aaronrobertshaw committed:
Improve color type descriptions

aaronrobertshaw committed:
Omit as prop from WordPressComponentProps

aaronrobertshaw committed:
Remove unnecessary backticks from type default

aaronrobertshaw committed:
Expand type definitions in RangeControlProps for better storybook docs

aaronrobertshaw committed:
Fix tooltip regression when step is any

mirka commented on packages/components/src/range-control/types.ts:
```diff
@@ -229,6 +230,11 @@ export type RangeControlProps< IconProps = unknown > = Pick<
+export type UnforwardedRangeControlProps< IconProps = unknown > = Omit<
+	WordPressComponentProps< RangeControlProps< IconProps >, 'div' >,
+	'as'
+>;
```
Just checking in case this is an oversight: `WordPressComponentProps` takes a boolean [`isPolymorphic` as its third argument](https://github.com/WordPress/gutenberg/blob/a63cb7b48017f6eeba3fd6c1ebf0161a7c70d5ad/packages/components/src/ui/context/wordpress-component.ts#L16) that basically handles this `as` prop business. Could we use that instead?

aaronrobertshaw replied:
Thanks @mirka. It was an oversight sorry. I've updated this to use the boolean argument.

aaronrobertshaw committed:
Fix removal of polymorphism for RangeControl

ciampo commented:
Given that #41444 and #41445 are approved and ready to merge, should we merge them into this PR and have a final review, or should we merge this PR first, and then rebase+merge #41444 and #41445 into `trunk` later?

aaronrobertshaw commented:
I'm happy to merge this and juggle the others if it saves others re-reviewing. I'll be back from some unexpected AFK next week and will get this sorted then.

ciampo commented:
Sure! This PR needs a rebase anyway — we can give a final review to this PR once you're back next week

aaronrobertshaw committed:
Fix post-rebase typing errors

aaronrobertshaw commented:
After rebasing this PR I encountered a few new typing errors. It appears there is a type conflict between the RangeControl and the ColorPicker which uses RangeControl internally for the RGB/HSL inputs.

The RangeControl needs to allow the `onChange` handler to accept param of type `number | undefined` whereas the ColorPicker inputs assume the new value is always a defined `number` and types its change handler accordingly. That change handler is also passed to the inner RangeControl. 

I've updated the ColorPicker to play nice but am not confident in the approach. I get the same behaviour now as on trunk in both Storybook and the editor for what that's worth.

@ciampo any thoughts on a better approach to the changes in https://github.com/WordPress/gutenberg/pull/40535/commits/a3046155be30de57a203a2f018b56052f90ec533?

kevin940726 commented:
> The RangeControl needs to allow the `onChange` handler to accept param of type `number | undefined`

Why does it need to allow `onChange` handler to accept `undefined`? 🤔

Sorry if I'm missing some context here 😅.

aaronrobertshaw commented:
Thanks for looking at this one @kevin940726 👍 


> Why does it need to allow onChange handler to accept undefined? 🤔

I don't have the full context or history on this one either however I believe its to do with how the reset handler was implemented originally. In the snippet below `onChange` can potentially receive an undefined value.

https://github.com/WordPress/gutenberg/blob/a3046155be30de57a203a2f018b56052f90ec533/packages/components/src/range-control/index.tsx#L179-L204

ciampo commented:
> I don't have the full context or history on this one either however I believe its to do with how the reset handler was implemented originally. In the snippet below `onChange` can potentially receive an undefined value.

The assumption is correct, and it is confirmed by the description in the README:

https://github.com/WordPress/gutenberg/blob/a3046155be30de57a203a2f018b56052f90ec533/packages/components/src/range-control/README.md?plain=1#L246-L251

Additional context can be found in a comment in the `handleOnReset` comment:

https://github.com/WordPress/gutenberg/blob/a3046155be30de57a203a2f018b56052f90ec533/packages/components/src/range-control/index.tsx#L190-L202


This is not a simple situation. 

First of all, I don't think that the concept of "resetting" a range input to an `undefined` value exists in the vanilla `input` element (rendering `<input type="range" value="" />` (or even `<input type="range" />`) results in an input field holding a value of `50`).

Also, as the code comment above says, passing `undefined` triggers a switch between the controlled and uncontrolled version of the component.

A potential solution could be to use an empty string as the fallback reset value (instead of `undefined`), and therefore also pass it as the argument of `onChange`:

<details>

<summary>(something like this)</summary>

```diff
diff --git a/packages/components/src/range-control/index.tsx b/packages/components/src/range-control/index.tsx
index 21e96bd8d9..1d89f7be20 100644
--- a/packages/components/src/range-control/index.tsx
+++ b/packages/components/src/range-control/index.tsx
@@ -177,30 +177,14 @@ function UnforwardedRangeControl< IconProps = unknown >(
 	};
 
 	const handleOnReset = () => {
-		let resetValue: number | null = parseFloat( `${ resetFallbackValue }` );
-		let onChangeResetValue: number | undefined = resetValue;
+		const parsedResetFallbackValue = parseFloat(`${resetFallbackValue}`);
+		const resetValue = isNaN(parsedResetFallbackValue) ? '' : parsedResetFallbackValue;
 
-		if ( isNaN( resetValue ) ) {
-			resetValue = null;
-			onChangeResetValue = undefined;
-		}
+		// Will cause a TypeScript error
+		setValue(resetValue);
 
-		setValue( resetValue );
-
-		/**
-		 * Previously, this callback would always receive undefined as
-		 * an argument. This behavior is unexpected, specifically
-		 * when resetFallbackValue is defined.
-		 *
-		 * The value of undefined is not ideal. Passing it through
-		 * to internal <input /> elements would change it from a
-		 * controlled component to an uncontrolled component.
-		 *
-		 * For now, to minimize unexpected regressions, we're going to
-		 * preserve the undefined callback argument, except when a
-		 * resetFallbackValue is defined.
-		 */
-		onChange( onChangeResetValue );
+		// Will cause a TypeScript error
+		onChange(resetValue);
 	};
 
 	const handleShowTooltip = () => setShowTooltip( true );
```
</details>

The issue with this approach is that it clashes with the assumption that we made so far that `value` is of type `number`.

Probably a good solution could be:
- add `@ts-expect-error` comments in the `ColorPicker` component, and try to merge this PR with the least amount of runtime changes possible
- look into addressing this problem in a separate PR — which would likely mean to reconsider the type of the `value` for all input components across the package (see https://github.com/WordPress/gutenberg/pull/40535#issuecomment-1112049865 for more considerations)

What do folks think?

aaronrobertshaw commented:
Thanks for the insight and guidance @ciampo 👍 

> add @ts-expect-error comments in the ColorPicker component, and try to merge this PR with the least amount of runtime changes possible
> 
> ...
> 
> What do folks think?

I think this is a wise approach. It will help land the initial TypeScript conversion of the component and potentially allow some follow-on work to occur concurrently to reconsidering the type of `value`.

I'll get the changes re: `@ts-expect-error` sorted so this can be ready for re-review early next week.

aaronrobertshaw committed:
Minimise runtime changes via ts-expect-error

See: https://github.com/WordPress/gutenberg/pull/40535#issuecomment-1172418185 for more information on why we are ignoring this error. TL;DR we are looking to address some conflicts between input control types for their value props.

aaronrobertshaw commented:
I've reverted the changes to the ColorPicker's RGB and HSL inputs as well as added the `@ts-expect-error` to where the ColorPicker passes its `onChange` to its inner RangeControl. (https://github.com/WordPress/gutenberg/pull/40535/commits/a1fa7d35d98517301b8cf5c992a8afc78acffde9).

🤞 This should be ready for a final review given the issue with `value` props typing to be addressed in follow-ups.

ciampo commented on packages/components/src/range-control/README.md:
```diff
@@ -109,75 +109,100 @@ const MyRangeControl = () => {
+### `afterIcon`: `string|Function|WPComponent|null`
 
-If this property is added, a label will be generated using label property as the content.
+If this property is added, a DashIcon component will be rendered after the slider with the icon equal to afterIcon.
```
Nit: the `Icon` component can render more icon "types" than just `dashicon`s (also, I believe we're trying to move away from `dashicon`s in the repo)

```suggestion
If this property is added, an [Icon component](/packages/components/src/icon/README.md) will be rendered after the slider with the icon equal to afterIcon.
```

ciampo commented on packages/components/src/range-control/README.md:
```diff
@@ -109,75 +109,100 @@ const MyRangeControl = () => {
+### `beforeIcon`: `string|Function|WPComponent|null`
 
-If this property is added, a help text will be generated using help property as the content.
+If this property is added, a DashIcon component will be rendered before the slider with the icon equal to beforeIcon.
```
Same as previous comment:

```suggestion
If this property is added, an [Icon component](/packages/components/src/icon/README.md) will be rendered before the slider with the icon equal to beforeIcon.
```

ciampo commented on packages/components/src/range-control/types.ts:
```diff
@@ -0,0 +1,288 @@
+		 *
+		 * For more information on `IconType` see the Icon component:
+		 * /packages/components/src/icon/index.tsx
+		 */
```
Same re. dashicons

```suggestion
		/**
		 * If this property is added, an Icon component will be rendered
		 * after the slider with the icon equal to afterIcon.
		 *
		 * @see /packages/components/src/icon/index.tsx
		 */
```

ciampo commented on packages/components/src/range-control/types.ts:
```diff
@@ -0,0 +1,288 @@
+		 *
+		 * For more information on `IconType` see the Icon component:
+		 * /packages/components/src/icon/index.tsx
+		 */
```
Same re. dashicons

```suggestion
		/**
		 * If this property is added, an Icon component will be rendered
		 * before the slider with the icon equal to beforeIcon.
		 *
		 * @see /packages/components/src/icon/index.tsx
		 */
```

ciampo commented on packages/components/src/range-control/types.ts:
```diff
@@ -0,0 +1,288 @@
+		 * The value to revert to if the Reset button is clicked (enabled by
+		 * `allowReset`)
+		 */
+		resetFallbackValue?: number | string | null;
```
Since we established that (at least for now) `value` should be of type `number`, should we remove `string` and `null`  from the possible types of `resetFallbackValue` ?

```suggestion
		resetFallbackValue?: number;
```

(this change also improves the Storybook control from this prop, as it changes from "Set object" to "Set number")

ciampo commented on packages/components/src/range-control/types.ts:
```diff
@@ -0,0 +1,288 @@
+};
+
+export type WrapperProps = {
+	className?: string;
```
```suggestion
```

No need to specify the `classname` prop here, as it's already added when using the `styled.div` type in 

https://github.com/WordPress/gutenberg/blob/307c26b24bebf1ec1c4644265750a7bdbeef64b5/packages/components/src/range-control/styles/range-control-styles.ts#L47

ciampo commented on packages/components/src/range-control/index.tsx:
```diff
@@ -35,67 +35,79 @@ import {
+function UnforwardedRangeControl< IconProps = unknown >(
+	props: WordPressComponentProps<
+		RangeControlProps< IconProps >,
+		'div',
```
This is an interesting consequence of how we type and forward refs in this component — normally, I'd expect this component to inherit props from the HTML `input` element (like, for example, [`UnitControl` does](https://github.com/WordPress/gutenberg/blob/33d84b036592a5bf31af05b7710f3b2b14163dc4/packages/components/src/unit-control/index.tsx#L40-L43))

@mirka , what you do think?

mirka replied:
If I'm following the code correctly, it _does_ look like it's supposed to inherit from `input`. That's where the `otherProps` ultimately go.

```suggestion
		'input',
```

(It's funny because `div` and `span` are perhaps the only two mistakes you can make in this context without TS throwing an error 😂)

ciampo commented on packages/components/src/range-control/index.tsx:
```diff
@@ -35,67 +35,79 @@ import {
+		'div',
+		false
+	>,
+	ref: ForwardedRef< HTMLDivElement >
```
Further down the code, It looks like `ref` is being passed an `HTMLInputElement`:

https://github.com/WordPress/gutenberg/blob/59f61d0689e1c00114fb724fc3f0a9c0802af90b/packages/components/src/range-control/index.tsx#L107-L113

Therefore, shouldn't `ref` be typed as a `ForwardedRef< HTMLInputElement >` too? I'm personally a bit confused to why TypeScript is not showing an error here

mirka replied:
Wait, isn't it supposed to be this 👇 to begin with?

```suggestion
	ref: ForwardedRef< HTMLInputElement >
```

---

> I'm personally a bit confused to why TypeScript is not showing an error here

```ts
type ForwardedRef<T> = ((instance: T | null) => void) | MutableRefObject<T | null> | null;

interface MutableRefObject<T> {
  current: T;
}
```

I guess it _is_ valid if it matches against `MutableRefObject< HTMLDivElement>` (instead of `(instance: HTMLDivElement) => void`)?

```js
// 👍 valid
const whateverFunc: React.MutableRefObject< HTMLDivElement > = ( myInput: HTMLInputElement ) => {};
```

If so, that kind of means the `React.ForwardedRef<T>` generic is mostly for type safety of methods on `ref.current` (e.g. `ref.current.focus()`), and not particularly useful for checking arg types in `ref(foo)` patterns.

ciampo commented on packages/components/src/range-control/index.tsx:
```diff
@@ -35,67 +35,79 @@ import {
 		inputRef.current = nodeRef;
 
-		if ( ref ) {
+		if ( typeof ref === 'function' ) {
```
Looking at the way `ref` is being invoked here, this means that only legacy, callback-style refs will work with `RangeControl`:

```tsx
// Legacy callback refs work as expected ✅
const RangeControlCallbackRef = ( props ) => {
	const [ value, setValue ] = useState();
	const ref = ( element ) => {
		// Outputs the `input` element correctly
		console.log( element );
	};

	return (
		<RangeControl
			{ ...props }
			value={ value }
			onChange={ setValue }
			ref={ ref }
		/>
	);
};

// Modern refs don't work ⛔️
const RangeControlModernRef = ( props ) => {
	const [ value, setValue ] = useState();

	const ref = useRef( null );
	// Always outputs `null`
	console.log( ref );

	return (
		<RangeControl
			{ ...props }
			value={ value }
			onChange={ setValue }
			ref={ ref }
		/>
	);
};
```

We could look into rewriting this part of the component and use the `useMergeRefs` hook:

<details>

<summary>Something like this</summary>

```diff
diff --git a/packages/components/src/range-control/index.tsx b/packages/components/src/range-control/index.tsx
index 21e96bd8d9..e4ac644b2e 100644
--- a/packages/components/src/range-control/index.tsx
+++ b/packages/components/src/range-control/index.tsx
@@ -9,7 +9,7 @@ import type { ChangeEvent, FocusEvent, ForwardedRef } from 'react';
  */
 import { __, isRTL } from '@wordpress/i18n';
 import { useRef, useState, forwardRef } from '@wordpress/element';
-import { useInstanceId } from '@wordpress/compose';
+import { useInstanceId, useMergeRefs } from '@wordpress/compose';
 
 /**
  * Internal dependencies
@@ -46,7 +46,7 @@ function UnforwardedRangeControl< IconProps = unknown >(
 		'div',
 		false
 	>,
-	ref: ForwardedRef< HTMLDivElement >
+	forwardedRef: ForwardedRef< HTMLDivElement >
 ) {
 	const {
 		afterIcon,
@@ -104,14 +104,6 @@ function UnforwardedRangeControl< IconProps = unknown >(
 
 	const inputRef = useRef< HTMLInputElement >();
 
-	const setRef = ( nodeRef: HTMLInputElement ) => {
-		inputRef.current = nodeRef;
-
-		if ( typeof ref === 'function' ) {
-			ref( nodeRef );
-		}
-	};
-
 	const isCurrentlyFocused = inputRef.current?.matches( ':focus' );
 	const isThumbFocused = ! disabled && isFocused;
 
@@ -255,7 +247,7 @@ function UnforwardedRangeControl< IconProps = unknown >(
 						onFocus={ handleOnFocus }
 						onMouseMove={ onMouseMove }
 						onMouseLeave={ onMouseLeave }
-						ref={ setRef }
+						ref={ useMergeRefs( [ inputRef, forwardedRef ] ) }
 						step={ step }
 						value={ inputSliderValue ?? undefined }
 					/>
```

</details>

ciampo commented on packages/components/src/range-control/rail.tsx:
```diff
@@ -29,24 +40,24 @@ export default function RangeRail( {
 					max={ max }
 					step={ step }
-					value={ value }
+					value={ value ?? ( max - min ) / 2 + min }
```
Is there a specific reason for why this fallback value was added here? (apologies in case this has already been explained)

stokesman replied:
🤔 My guess is it’s an artifact of rebasing (from when this was based on #40568).

ciampo commented on packages/components/src/range-control/types.ts:
```diff
@@ -0,0 +1,288 @@
+};
+
+export type ControlledRangeValue = number | '' | null;
+export type SetControlledRangeValue = ( value?: number | null ) => void;
```
This type doesn't seem to be used anywhere, can it be deleted?

ciampo commented on packages/components/src/range-control/types.ts:
```diff
@@ -0,0 +1,288 @@
+export type InputRangeProps = {
+	describedBy?: string;
+	label?: string;
+	onHideTooltip?: () => undefined;
```
Unless we specifically want to flag that this function returns `undefined`, we should use the `void` type for functions whose return value should not be observed

```suggestion
	onHideTooltip?: () => void;
```

ciampo commented on packages/components/src/range-control/types.ts:
```diff
@@ -0,0 +1,288 @@
+	onHideTooltip?: () => undefined;
+	onMouseLeave?: MouseEventHandler< HTMLInputElement >;
+	onMouseMove?: MouseEventHandler< HTMLInputElement >;
+	onShowTooltip?: () => undefined;
```
Same re. `undefined` vs `void` for function return values

```suggestion
	onShowTooltip?: () => void;
```

ciampo commented on packages/components/src/range-control/utils.ts:
```diff
@@ -20,24 +28,24 @@ const noop = () => {};
  * @param {number}      max   The maximum value.
  *
- * @return {number} A (float) number
+ * @return {number | null} A (float) number
```
Since we're adding types via TypeScript, they can be removed from the JsDoc annotations

<details>

<summary>Something like this</summary>

```diff
diff --git a/packages/components/src/range-control/utils.ts b/packages/components/src/range-control/utils.ts
index 7a455134a1..3eb2cf86cd 100644
--- a/packages/components/src/range-control/utils.ts
+++ b/packages/components/src/range-control/utils.ts
@@ -24,11 +24,11 @@ const noop = () => {};
 /**
  * A float supported clamp function for a specific value.
  *
- * @param {number|null} value The value to clamp.
- * @param {number}      min   The minimum value.
- * @param {number}      max   The maximum value.
+ * @param  value The value to clamp.
+ * @param  min   The minimum value.
+ * @param  max   The maximum value.
  *
- * @return {number | null} A (float) number
+ * @return A (float) number
  */
 export function floatClamp( value: number | null, min: number, max: number ) {
 	if ( typeof value !== 'number' ) {
@@ -41,13 +41,13 @@ export function floatClamp( value: number | null, min: number, max: number ) {
 /**
  * Hook to store a clamped value, derived from props.
  *
- * @param {Object}        settings         Hook settings.
- * @param {number}        settings.min     The minimum value.
- * @param {number}        settings.max     The maximum value.
- * @param {number | null} settings.value   The current value.
- * @param {any}           settings.initial The initial value.
+ * @param  settings         Hook settings.
+ * @param  settings.min     The minimum value.
+ * @param  settings.max     The maximum value.
+ * @param  settings.value   The current value.
+ * @param  settings.initial The initial value.
  *
- * @return {[*, Function]} The controlled value and the value setter.
+ * @return The controlled value and the value setter.
  */
 export function useControlledRangeValue( {
 	min,
@@ -78,14 +78,14 @@ export function useControlledRangeValue( {
  * Hook to encapsulate the debouncing "hover" to better handle the showing
  * and hiding of the Tooltip.
  *
- * @param {Object}   settings                     Hook settings.
- * @param {Function} [settings.onShow=noop]       A callback function invoked when the element is shown.
- * @param {Function} [settings.onHide=noop]       A callback function invoked when the element is hidden.
- * @param {Function} [settings.onMouseMove=noop]  A callback function invoked when the mouse is moved.
- * @param {Function} [settings.onMouseLeave=noop] A callback function invoked when the mouse is moved out of the element.
- * @param {number}   [settings.timeout=300]       Timeout before the element is shown or hidden.
+ * @param  settings                     Hook settings.
+ * @param  [settings.onShow=noop]       A callback function invoked when the element is shown.
+ * @param  [settings.onHide=noop]       A callback function invoked when the element is hidden.
+ * @param  [settings.onMouseMove=noop]  A callback function invoked when the mouse is moved.
+ * @param  [settings.onMouseLeave=noop] A callback function invoked when the mouse is moved out of the element.
+ * @param  [settings.timeout=300]       Timeout before the element is shown or hidden.
  *
- * @return {Object} Bound properties for use on a React.Node.
+ * @return Bound properties for use on a React.Node.
  */
 export function useDebouncedHoverInteraction( {
 	onHide = noop,
```
</details>

@sarayourfriend, would you also remove from the JSDocs other notation (like default values and the square brackets to mark props as optional) ?

mirka replied:
> would you also remove from the JSDocs other notation (like default values and the square brackets to mark props as optional) ?

If anything, I think it's beneficial to move those out of the top-level function JSDoc and into "inline" JSDocs:

```js
/**
 * Hook to encapsulate the debouncing "hover" to better handle the showing
 * and hiding of the Tooltip.
 *
 * @return Bound properties for use on a React.Node.
 */
export function useDebouncedHoverInteraction( {
	onHide = noop,
	onMouseLeave = noop as MouseEventHandler,
	onMouseMove = noop as MouseEventHandler,
	onShow = noop,
	timeout = 300,
}: {
	/**
	 * A callback function invoked when the element is hidden.
	 *
	 * @default noop
	 */
	onHide?: () => void;
	/* etc... */
} ) {
```

This gives you better availability of the docs via IntelliSense — when writing out an invocation of the function, and also when hovering over the parameters in the internal function code. (At least in VS Code. YMMV.)

ciampo commented on packages/components/src/range-control/utils.ts:
```diff
@@ -46,7 +54,7 @@ export function useControlledRangeValue( {
 	value: valueProp,
 	initial,
-} ) {
+}: useControlledRangeValueProps ): useControlledRangeValueReturn {
```
This is a nit, so feel free to ignore — but in general, I prefer to use as least hardcoded types as possible, and let the TypeScript engine infer the rest.

In this case, we could avoid the return type and rewrite the code like this

<details>

<summary>Click to expand</summary>

```diff
diff --git a/packages/components/src/range-control/types.ts b/packages/components/src/range-control/types.ts
index 0fa4c45118..eda285376a 100644
--- a/packages/components/src/range-control/types.ts
+++ b/packages/components/src/range-control/types.ts
@@ -270,18 +270,13 @@ export type TrackProps = {
 	trackColor: CSSProperties[ 'color' ];
 };
 
-export type useControlledRangeValueProps = {
+export type useControlledRangeValueArgs = {
 	initial?: number;
 	max: number;
 	min: number;
 	value: number | null;
 };
 
-export type useControlledRangeValueReturn = [
-	number | '' | null,
-	( nextValue: number | null ) => void
-];
-
 export type useMarksArgs = NumericProps & {
 	marks: RangeMarks;
 	step: number;
diff --git a/packages/components/src/range-control/utils.ts b/packages/components/src/range-control/utils.ts
index 7a455134a1..2e55db6f7d 100644
--- a/packages/components/src/range-control/utils.ts
+++ b/packages/components/src/range-control/utils.ts
@@ -14,10 +14,7 @@ import { useCallback, useRef, useEffect, useState } from '@wordpress/element';
 import { useControlledState } from '../utils/hooks';
 import { clamp } from '../utils/math';
 
-import type {
-	useControlledRangeValueProps,
-	useControlledRangeValueReturn,
-} from './types';
+import type { useControlledRangeValueArgs } from './types';
 
 const noop = () => {};
 
@@ -54,14 +51,14 @@ export function useControlledRangeValue( {
 	max,
 	value: valueProp,
 	initial,
-}: useControlledRangeValueProps ): useControlledRangeValueReturn {
+}: useControlledRangeValueArgs ) {
 	const [ state, setInternalState ] = useControlledState(
 		floatClamp( valueProp, min, max ),
 		{ initial, fallback: null }
 	);
 
 	const setState = useCallback(
-		( nextValue ) => {
+		( nextValue: number | null ) => {
 			if ( nextValue === null ) {
 				setInternalState( null );
 			} else {
@@ -71,7 +68,7 @@ export function useControlledRangeValue( {
 		[ min, max ]
 	);
 
-	return [ state, setState ];
+	return [ state, setState ] as const;
 }
 
 /**
```
</details>

ciampo reviewed (commented):
> I've reverted the changes to the ColorPicker's RGB and HSL inputs as well as added the @ts-expect-error to where the ColorPicker passes its onChange to its inner RangeControl. 

Thank you Aaron, the changes to `ColorPicker` are reduced to a minimun!

Today I spent a good amount of time and took an in-depth look at the code. I added a few more comments, although they're all quite small observations (sorry if this is causing one additional iteration on this PR 🙏 )

----

During my tests, I also noticed an unexpected behavior of the component: when resetting the component's value (or before the user interacts with the component, the value of the `initialPosition` is not clamped according to `min` and `max` :

https://user-images.githubusercontent.com/1083581/177184039-54403071-1e2c-4aa5-a8f3-8333127c9119.mp4

Although this behaviour can also be reproduced on `trunk`, and therefore I believe we should work on a fix separately from this PR.

aaronrobertshaw committed:
Clarify icon prop descriptions

aaronrobertshaw committed:
Revert resetFallbackValue type to number

aaronrobertshaw committed:
Remove unrequired className from WrapperProps

aaronrobertshaw committed:
Fix WordPressComponentProps and ForwardedRef types

aaronrobertshaw committed:
Leverage useMergeRefs to fix for modern refs

aaronrobertshaw committed:
Fix rebase error

aaronrobertshaw committed:
Remove unused SetControlledRangeValue type

aaronrobertshaw committed:
Fix onShowTooltip and onHideTooltip types

aaronrobertshaw committed:
Remove types from JSDocs

aaronrobertshaw committed:
Improve useControlledRangeValue hook types

aaronrobertshaw commented:
Thanks for the in-depth review @ciampo and the extra insights @mirka 🙇 

I've cleaned up this PR with the suggested improvements and fixes. The only suggestion I wasn't too sure about was the [potential change to the jsdocs](https://github.com/WordPress/gutenberg/pull/40535#discussion_r913256248) for the util functions. 

Moving the prop descriptions etc to "inline" JSDocs while still leaving the top-level function JSDoc comment gives me linting errors about missing params. What's our preferred result here?

ciampo commented on packages/components/src/range-control/types.ts:
```diff
@@ -0,0 +1,281 @@
+		 */
+		currentInput?: number;
+		/**
+		 * An icon to be shown above the slider next to it's container title.
```
```suggestion
		 * An icon to be shown above the slider next to its container title.
```

ciampo commented on packages/components/src/range-control/README.md:
```diff
@@ -109,75 +109,100 @@ const MyRangeControl = () => {
+
+### `icon`: `string`
+
+An icon to be shown above the slider next to it's container title.
```
```suggestion
An icon to be shown above the slider next to its container title.
```

mirka commented:
> Moving the prop descriptions etc to "inline" JSDocs while still leaving the top-level function JSDoc comment gives me linting errors about missing params. What's our preferred result here?

Ugh, that thing 😬 The rules are [a little bit lenient](https://github.com/WordPress/gutenberg/blob/64a42a25c028cf8c2c9791fe255f5028167d8ccf/.eslintrc.js#L330-L340) in the components package since we have specific docgen needs, but we started by only including `tsx` files (i.e. component files) and not `ts` files (i.e. utils), to be conservative. I didn't realize at the time that inline JSDocs were not just better for our Storybook docgen, but also for the VS Code IntelliSense. Though, because we don't have strong conventions for type data in our `ts` files, I'm not confident that disabling `jsdoc/require-param` for `ts` files won't have a negative effect on our documentation coverage. Maybe we'll revisit that and add `ts` files to the eslint override at some point, but for now perhaps we can just list the minimum viable info to satisfy the linter:

```js
/**
 * Hook to encapsulate the debouncing "hover" to better handle the showing
 * and hiding of the Tooltip.
 *
 * @param  settings
 * @param  settings.onShow
 * @param  settings.onHide
 * @param  settings.onMouseMove
 * @param  settings.onMouseLeave
 * @param  settings.timeout
 *
 * @return Bound properties for use on a React.Node.
 */
export function useDebouncedHoverInteraction( {
```

(If you want, it's also possible to move the arg destructuring into the function body, and then you'll only have to list `@param settings` in the JSDoc.)

ciampo commented on packages/components/src/range-control/index.tsx:
```diff
@@ -35,82 +35,88 @@ import {
-
-	const calculatedFillValue = ( ( value - min ) / ( max - min ) ) * 100;
+	const calculatedFillValue =
+		( ( rangeFillValue - min ) / ( max - min ) ) * 100;
```
We can actually avoid introducing runtime changes here with a couple of small tweaks:
- we can cast the type of `state` in `useControlledState` to signal that it can't be an empty string (that's something that could be improved in `useControlledState`'s types)
- we can inline the computation of `calculatedFillValue` into the next line (since `isValueReset` acts as a type guard against `value` being `null`)

```diff
diff --git a/packages/components/src/range-control/index.tsx b/packages/components/src/range-control/index.tsx
index 23144cf0e9..7f5a7c4ae3 100644
--- a/packages/components/src/range-control/index.tsx
+++ b/packages/components/src/range-control/index.tsx
@@ -110,14 +110,9 @@ function UnforwardedRangeControl< IconProps = unknown >(
 	const currentValue = value !== undefined ? value : currentInput;
 
 	const inputSliderValue = isValueReset ? '' : currentValue;
-	const numericValue = parseFloat( `${ value }` );
-	const rangeFillValue = isValueReset
-		? ( max - min ) / 2 + min
-		: numericValue;
-
-	const calculatedFillValue =
-		( ( rangeFillValue - min ) / ( max - min ) ) * 100;
-	const fillValue = isValueReset ? 50 : calculatedFillValue;
+	const rangeFillValue = isValueReset ? ( max - min ) / 2 + min : value;
+
+	const fillValue = isValueReset ? 50 : ( ( value - min ) / ( max - min ) ) * 100;
 	const fillValueOffset = `${ clamp( fillValue, 0, 100 ) }%`;
 
 	const classes = classnames( 'components-range-control', className );
diff --git a/packages/components/src/range-control/utils.ts b/packages/components/src/range-control/utils.ts
index 053c20f12f..a8bc8b99b5 100644
--- a/packages/components/src/range-control/utils.ts
+++ b/packages/components/src/range-control/utils.ts
@@ -68,7 +68,9 @@ export function useControlledRangeValue( {
 		[ min, max ]
 	);
 
-	return [ state, setState ] as const;
+	// `state` can't be an empty string because we specified a fallback value of
+	// `null` in `useControlledState`
+	return [ state as Exclude<typeof state, ''>, setState ] as const;
 }
 
 /**
```

aaronrobertshaw replied:
Nice. Thank you for the diff demonstrating the changes, appreciate it!

I've applied these and re-tested. Works well for me.

ciampo reviewed (commented):
> Moving the prop descriptions etc to "inline" JSDocs while still leaving the top-level function JSDoc comment gives me linting errors about missing params. What's our preferred result here?

My original comment was mostly about avoiding duplicated types (i.e. in the top-level JSDocs and in TypeScript) — otherwise I trust @mirka's opinion, as she's got more experience dealing with the interaction between typescript and JSDocs while working with Storybook's docgen.

aaronrobertshaw committed:
Fix copy and pasted typo

aaronrobertshaw committed:
Apply suggested feedback to remove runtime changes

aaronrobertshaw committed:
Tweak JSDoc comments and typing

aaronrobertshaw commented:
I'm still not 100% clear on what the best approach to the util function JSDoc comments is.

I've moved the destructuring of the settings arg for `useControlledRangeValue` and `useDebouncedHoverInteraction` into their function bodies. If I define the type inline for the `settings` arg, the linter still wants the `settings` object broken down in the JSDoc comment. If I define the type in `types.ts` and use that, it doesn't object but it appears that the intellisense results aren't as helpful if hovering the util function name. Once you start creating the settings object when calling the util function you get the helpful suggestions/info though.

These changes were pushed in https://github.com/WordPress/gutenberg/pull/40535/commits/a435f8140a94f8e888465cd20e48c0d27a338261 but I'm happy to revert that if I've gone off track. Any further wisdom to share here @mirka would be appreciated 🙇

mirka commented on packages/components/src/range-control/types.ts:
```diff
@@ -0,0 +1,326 @@
+	trackColor: CSSProperties[ 'color' ];
+};
+
+export type useControlledRangeValueArgs = {
```
These should start with uppercase, no?

```suggestion
export type UseControlledRangeValueArgs = {
```

aaronrobertshaw replied:
Thanks, I've fixed up the casing for the custom hook types.

mirka reviewed (approved):
> I'm still not 100% clear on what the best approach to the util function JSDoc comments is.

You're right, it's a bummer there isn't a perfect solution 🙁 I don't feel strongly about the specifics at the moment, as long as the type or description data is not duplicated. Whatever feels better in each context.

Looks like all feedback has been addressed. I noticed one minor thing about the type alias capitalization, but aside from that I think we're good to go 🚀

aaronrobertshaw committed:
Fix case for custom hook types

aaronrobertshaw commented:
Thanks for all the help and guidance on getting the RangeControl refactored @ciampo & @mirka, appreciate it 🙇 

This PR as well as the updates to the component's storybook example and unit tests have been merged.
